lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -81,16 +81,18 @@ def fragment_cache(cached_hash, non_cached_hash)
 
         def serializable_hash_for_collection(options)
           hash = { data: [] }
+          included = []
           serializer.each do |s|
             result = self.class.new(s, instance_options.merge(fieldset: fieldset)).serializable_hash(options)
             hash[:data] << result[:data]
+            next unless result[:included]
 
-            if result[:included]
-              hash[:included] ||= []
-              hash[:included] |= result[:included]
-            end
+            included |= result[:included]
           end
 
+          included.delete_if { |resource| hash[:data].include?(resource) }
```

Comment:
It is a structural issue. When the primary resource is a single resource (`serializable_hash_for_single_resource`), we make sure we do not add to `included` any related resource that is the primary resource. However, when the primary resource is a collection (`serializable_hash_for_collection`), the adapter works by computing the hash for every resource independently, and then combining the hashes into a big one. This has the advantage of clear code, but the disadvantage that when serializing `[a, b]` where `a` includes `b`, we have no other choice than manually removing duplicates when aggregating.